File: src/api/parse.ts
import { History, PricePoint, Quote, RangeKey, SearchResult } from './types';

/**
 * Pure parsers for the Yahoo Finance JSON payloads.
 *
 * These are deliberately kept free of `fetch` so they can be unit tested
 * against recorded fixtures. The upstream feed is undocumented and changes
 * shape occasionally, so every field is read defensively: a missing or
 * non-numeric field degrades the result rather than throwing.
 */

/** Thrown when a payload is structurally unusable (no result, or upstream error). */
export class FeedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FeedError';
  }
}

function num(v: unknown): number | null {
  return typeof v === 'number' && Number.isFinite(v) ? v : null;
}

function str(v: unknown): string | null {
  return typeof v === 'string' && v.length > 0 ? v : null;
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null;
}

/** Pulls `chart.result[0]`, raising a FeedError if the payload carries one. */
function chartResult(payload: unknown): Record<string, unknown> {
  if (!isRecord(payload) || !isRecord(payload.chart)) {
    throw new FeedError('Unexpected response: no chart object');
  }
  const chart = payload.chart;
  const err = chart.error;
  if (isRecord(err)) {
    throw new FeedError(str(err.description) ?? str(err.code) ?? 'Upstream error');
  }
  const results = chart.result;
  if (!Array.isArray(results) || !isRecord(results[0])) {
    throw new FeedError('Unexpected response: no chart data');
  }
  return results[0];
}

/**
 * Zips the parallel `timestamp` and `indicators.quote[0].close` arrays into
 * points, dropping entries where either side is null. Yahoo pads both arrays
 * with nulls for halted or not-yet-traded intervals.
 */
export function parsePoints(result: Record<string, unknown>): PricePoint[] {
  const timestamps = Array.isArray(result.timestamp) ? result.timestamp : [];
  const indicators = isRecord(result.indicators) ? result.indicators : {};
  const quoteArr = Array.isArray(indicators.quote) ? indicators.quote : [];
  const quote0 = isRecord(quoteArr[0]) ? quoteArr[0] : {};
  const closes = Array.isArray(quote0.close) ? quote0.close : [];

  const points: PricePoint[] = [];
  const n = Math.min(timestamps.length, closes.length);
  for (let i = 0; i < n; i++) {
    const t = num(timestamps[i]);
    const c = num(closes[i]);
    if (t !== null && c !== null) points.push({ t, c });
  }
  return points;
}

/** Builds a watchlist quote from a `range=1d` chart payload. */
export function parseQuote(payload: unknown, fetchedAt = Date.now()): Quote {
  const result = chartResult(payload);
  const meta = isRecord(result.meta) ? result.meta : {};
  const points = parsePoints(result);
  const lastPoint = points.length > 0 ? points[points.length - 1].c : null;

  const symbol = str(meta.symbol) ?? '';
  if (!symbol) throw new FeedError('Unexpected response: no symbol');

  const price = num(meta.regularMarketPrice) ?? lastPoint;
  if (price === null) throw new FeedError(`No price available for ${symbol}`);

  // previousClose is the prior session's close; chartPreviousClose is the close
  // before the requested range begins. For a 1d range they agree, but only the
  // former is present on every payload.
  const previousClose =
    num(meta.previousClose) ??
    num(meta.chartPreviousClose) ??
    (points.length > 0 ? points[0].c : price);

  const change = price - previousClose;

  return {
    symbol,
    name: str(meta.longName) ?? str(meta.shortName) ?? symbol,
    price,
    previousClose,
    change,
    // A zero previous close would only happen on a malformed payload; guard
    // anyway so the UI never renders NaN%.
    changePercent: previousClose !== 0 ? (change / previousClose) * 100 : 0,
    currency: str(meta.currency) ?? 'USD',
    exchange: str(meta.fullExchangeName) ?? str(meta.exchangeName) ?? '',
    marketState: str(meta.marketState) ?? '',
    dayHigh: num(meta.regularMarketDayHigh),
    dayLow: num(meta.regularMarketDayLow),
    spark: points,
    fetchedAt,
  };
}

/** Builds a detail-screen series from a chart payload for the given range. */
export function parseHistory(payload: unknown, range: RangeKey): History {
  const result = chartResult(payload);
  const meta = isRecord(result.meta) ? result.meta : {};
  const points = parsePoints(result);

  if (points.length === 0) {
    throw new FeedError('No price history available for this range');
  }

  // For an intraday range the day's move is measured from the previous close,
  // not from the first tick of the session — otherwise an overnight gap
  // silently disappears from the chart's headline number.
  const baseline =
    range === '1D'
      ? (num(meta.chartPreviousClose) ?? num(meta.previousClose) ?? points[0].c)
      : points[0].c;

  const last = points[points.length - 1].c;
  const change = last - baseline;

  return {
    symbol: str(meta.symbol) ?? '',
    range,
    points,
    currency: str(meta.currency) ?? 'USD',
    first: baseline,
    last,
    change,
    changePercent: baseline !== 0 ? (change / baseline) * 100 : 0,
  };
}

/** Symbol types worth showing in search. Yahoo also returns futures, options
 *  and non-tradable entries that this app has nothing useful to do with. */
const SEARCHABLE_TYPES = new Set(['EQUITY', 'ETF', 'MUTUALFUND', 'INDEX', 'CRYPTOCURRENCY', 'CURRENCY']);

/** Parses the symbol lookup payload, dropping non-security hits. */
export function parseSearch(payload: unknown): SearchResult[] {
  if (!isRecord(payload) || !Array.isArray(payload.quotes)) return [];

  const out: SearchResult[] = [];
  for (const raw of payload.quotes) {
    if (!isRecord(raw)) continue;
    const symbol = str(raw.symbol);
    if (!symbol) continue;
    const type = (str(raw.quoteType) ?? '').toUpperCase();
    if (!SEARCHABLE_TYPES.has(type)) continue;
    out.push({
      symbol,
      name: str(raw.longname) ?? str(raw.shortname) ?? symbol,
      exchange: str(raw.exchDisp) ?? str(raw.exchange) ?? '',
      type,
    });
  }
  return out;
}

File: lib/api/parse.dart
import '../models/types.dart';

/// Pure parsers for the Yahoo Finance JSON payloads.
///
/// These are deliberately kept free of any HTTP so they can be unit tested
/// against recorded fixtures. The upstream feed is undocumented and changes
/// shape occasionally, so every field is read defensively: a missing or
/// non-numeric field degrades the result rather than throwing.

/// Thrown when a payload is structurally unusable (no result, or upstream
/// error).
class FeedException implements Exception {
  const FeedException(this.message);

  final String message;

  @override
  String toString() => message;
}

double? _num(Object? v) {
  if (v is num && v.isFinite) return v.toDouble();
  return null;
}

String? _str(Object? v) {
  if (v is String && v.isNotEmpty) return v;
  return null;
}

Map<String, dynamic>? _map(Object? v) {
  if (v is Map<String, dynamic>) return v;
  if (v is Map) return v.cast<String, dynamic>();
  return null;
}

List<dynamic> _list(Object? v) => v is List ? v : const [];

/// Pulls `chart.result[0]`, raising a [FeedException] if the payload carries an
/// error object instead.
Map<String, dynamic> _chartResult(Object? payload) {
  final root = _map(payload);
  final chart = root == null ? null : _map(root['chart']);
  if (chart == null) {
    throw const FeedException('Unexpected response: no chart object');
  }

  final err = _map(chart['error']);
  if (err != null) {
    throw FeedException(
      _str(err['description']) ?? _str(err['code']) ?? 'Upstream error',
    );
  }

  final results = _list(chart['result']);
  final first = results.isEmpty ? null : _map(results.first);
  if (first == null) {
    throw const FeedException('Unexpected response: no chart data');
  }
  return first;
}

/// Zips the parallel `timestamp` and `indicators.quote[0].close` arrays into
/// points, dropping entries where either side is null. Yahoo pads both arrays
/// with nulls for halted or not-yet-traded intervals.
List<PricePoint> parsePoints(Map<String, dynamic> result) {
  final timestamps = _list(result['timestamp']);
  final indicators = _map(result['indicators']) ?? const <String, dynamic>{};
  final quoteArr = _list(indicators['quote']);
  final quote0 = quoteArr.isEmpty
      ? null
      : _map(quoteArr.first) ?? const <String, dynamic>{};
  final closes = _list(quote0?['close']);

  final points = <PricePoint>[];
  final n = timestamps.length < closes.length
      ? timestamps.length
      : closes.length;
  for (var i = 0; i < n; i++) {
    final t = _num(timestamps[i]);
    final c = _num(closes[i]);
    if (t != null && c != null) {
      points.add(PricePoint(t: t.toInt(), c: c));
    }
  }
  return points;
}

/// Builds a watchlist quote from a `range=1d` chart payload.
Quote parseQuote(Object? payload, {int? fetchedAt}) {
  final result = _chartResult(payload);
  final meta = _map(result['meta']) ?? const <String, dynamic>{};
  final points = parsePoints(result);
  final lastPoint = points.isNotEmpty ? points.last.c : null;

  final symbol = _str(meta['symbol']);
  if (symbol == null) {
    throw const FeedException('Unexpected response: no symbol');
  }

  final price = _num(meta['regularMarketPrice']) ?? lastPoint;
  if (price == null) {
    throw FeedException('No price available for $symbol');
  }

  // previousClose is the prior session's close; chartPreviousClose is the close
  // before the requested range begins. For a 1d range they agree, but only the
  // former is present on every payload.
  final previousClose =
      _num(meta['previousClose']) ??
      _num(meta['chartPreviousClose']) ??
      (points.isNotEmpty ? points.first.c : price);

  final change = price - previousClose;

  return Quote(
    symbol: symbol,
    name: _str(meta['longName']) ?? _str(meta['shortName']) ?? symbol,
    price: price,
    previousClose: previousClose,
    change: change,
    // A zero previous close would only happen on a malformed payload; guard
    // anyway so the UI never renders NaN%.
    changePercent: previousClose != 0 ? (change / previousClose) * 100 : 0,
    currency: _str(meta['currency']) ?? 'USD',
    exchange:
        _str(meta['fullExchangeName']) ?? _str(meta['exchangeName']) ?? '',
    marketState: _str(meta['marketState']) ?? '',
    dayHigh: _num(meta['regularMarketDayHigh']),
    dayLow: _num(meta['regularMarketDayLow']),
    spark: points,
    fetchedAt: fetchedAt ?? DateTime.now().millisecondsSinceEpoch,
  );
}

/// Builds a detail-screen series from a chart payload for the given range.
History parseHistory(Object? payload, RangeKey range) {
  final result = _chartResult(payload);
  final meta = _map(result['meta']) ?? const <String, dynamic>{};
  final points = parsePoints(result);

  if (points.isEmpty) {
    throw const FeedException('No price history available for this range');
  }

  // For an intraday range the day's move is measured from the previous close,
  // not from the first tick of the session — otherwise an overnight gap
  // silently disappears from the chart's headline number.
  final baseline = range == RangeKey.d1
      ? (_num(meta['chartPreviousClose']) ??
            _num(meta['previousClose']) ??
            points.first.c)
      : points.first.c;

  final last = points.last.c;
  final change = last - baseline;

  return History(
    symbol: _str(meta['symbol']) ?? '',
    range: range,
    points: points,
    currency: _str(meta['currency']) ?? 'USD',
    first: baseline,
    last: last,
    change: change,
    changePercent: baseline != 0 ? (change / baseline) * 100 : 0,
  );
}

/// Symbol types worth showing in search. Yahoo also returns futures, options
/// and non-tradable entries that this app has nothing useful to do with.
const _searchableTypes = <String>{
  'EQUITY',
  'ETF',
  'MUTUALFUND',
  'INDEX',
  'CRYPTOCURRENCY',
  'CURRENCY',
};

/// Parses the symbol lookup payload, dropping non-security hits.
List<SearchResult> parseSearch(Object? payload) {
  final root = _map(payload);
  if (root == null) return const [];

  final out = <SearchResult>[];
  for (final raw in _list(root['quotes'])) {
    final hit = _map(raw);
    if (hit == null) continue;
    final symbol = _str(hit['symbol']);
    if (symbol == null) continue;
    final type = (_str(hit['quoteType']) ?? '').toUpperCase();
    if (!_searchableTypes.contains(type)) continue;
    out.add(
      SearchResult(
        symbol: symbol,
        name: _str(hit['longname']) ?? _str(hit['shortname']) ?? symbol,
        exchange: _str(hit['exchDisp']) ?? _str(hit['exchange']) ?? '',
        type: type,
      ),
    );
  }
  return out;
}
